controlled addition

dobbse42 opened the pull request:
Following is a copy/paste from the email I sent you:

I have written the ctrl add function, which is the majority of the algorithm, and now only need to write the actual multiplier, which is essentially a few toffolis and a bunch of calls to the ctrl add function with different inputs.

Commit:
This is just to test that I can properly commit something. The actual code is basically meaningless

Commit:
added comments w/ pseudocode, ctrladd may or may not be deleted depending on what I decided regarding OOP

Commit:
Finished implementation of ctrladd, fixed various bugs in actual code (but not pseudocode), changed qubits generation, rewrote main, created changelog and to do.

dobbse42 committed:
was just a testing file

Commit:
Finished implementation of ctrladd, fixed various bugs in actual code (but not pseudocode), changed
        qubits generation, rewrote main, created changelog and to do.

Commit:
Merge branch 'master' of https://github.com/dobbse42/arithmetic

alexandrupaler commented on main.py:
```diff
@@ -0,0 +1,82 @@
+
+# def the following as ctrl_add with ctrl as the ctrl, qubitsSumnA as the bits of one summand, and qubitsSumnB as the
+# bits of the other, with the result stored in qubitsSumnB. qubitsSumnA[n] is ancillary, qubitsSumnB[n] = s[4] dot ctrl
+def ctrl_add (ctrl, qubitsSumnA, qubitsSumnB, n, circuit):      # circuit is possibly unnecessary, n is definitely unnecessary
```
make `circuit` the first parameter

alexandrupaler replied:
I think it would be best to infer `n` from `qubitsSumB` and not pass it as parameter

alexandrupaler commented on main.py:
```diff
@@ -0,0 +1,82 @@
+        qubits generation, rewrote main, created changelog and to do.
+    2/26/20: Wrote out ctrladd function in pseudocode, implemented through step 4
+    2/25/20: Wrote basic code to create qubits
+todo:
```
place these as issues in the repository. we could generate a "todo" tag

alexandrupaler commented on main.py:
```diff
@@ -0,0 +1,82 @@
+changelog:
+    3/3/20: Finished implementation of ctrladd, fixed various bugs in actual code (but not pseudocode), changed
+        qubits generation, rewrote main, created changelog and to do.
+    2/26/20: Wrote out ctrladd function in pseudocode, implemented through step 4
```
not necessary, because we have git, and the history is stored in the commit messages

Commit:
Changed n to be inferred from the len of qubitsSumnB, moved circuit to
bt the first parameter rather than last, and moved documentation at the
beginning into github.

Commit:
Added testMultiply to actually use the multiplier, as well as multiply
which functions correctly for multiplying two numbers of the same number
of bits.

Commit:
Moved everything to separate classes and wrote a more detailed example program (contained in the class main) giving options on how to test the code.
